bindings/c/src/types.rs

```diff
@@ -31,46 +31,55 @@ use opendal::Buffer;
 #[repr(C)]
 pub struct opendal_bytes {
     /// Pointing to the byte array on heap
-    pub data: *const u8,
+    pub data: *mut u8,
     /// The length of the byte array
     pub len: usize,
     /// The capacity of the byte array
     pub capacity: usize,
 }
 
 impl opendal_bytes {
+    pub(crate) fn empty() -> Self {
+        Self {
+            data: std::ptr::null_mut(),
+            len: 0,
+            capacity: 0,
+        }
+    }
+
     /// Construct a [`opendal_bytes`] from the Rust [`Vec`] of bytes
-    pub(crate) fn new(buf: Buffer) -> Self {
-        let vec = buf.to_vec();
-        let mut buf = std::mem::ManuallyDrop::new(vec);
-        let data = buf.as_mut_ptr();
-        let len = buf.len();
-        let capacity = buf.capacity();
+    pub(crate) fn new(b: Buffer) -> Self {
+        let mut b = std::mem::ManuallyDrop::new(b.to_vec());
         Self {
-            data,
-            len,
-            capacity,
+            data: b.as_mut_ptr(),
+            len: b.len(),
+            capacity: b.capacity(),
         }
     }
 
     /// \brief Frees the heap memory used by the opendal_bytes
     #[no_mangle]
-    pub unsafe extern "C" fn opendal_bytes_free(ptr: *mut opendal_bytes) {
-        if !ptr.is_null() {
-            // transmuting `*const u8` to `*mut u8` is undefined behavior in any cases
-            // however, fields type of `opendal_bytes` is already related to the zig binding
-            // it should be fixed later
-            let _ = Vec::from_raw_parts((*ptr).data as *mut u8, (*ptr).len, (*ptr).capacity);
-            // it is too weird that call `Box::new` outside `opendal_bytes::new` but dealloc it here
-            // also, boxing `opendal_bytes` might not be necessary
-            // `data` points to heap, so `opendal_bytes` could be passed as a stack value
-            let _ = Box::from_raw(ptr);
+    pub unsafe extern "C" fn opendal_bytes_free(bs: opendal_bytes) {
```

Comment:
Perhaps this method should still take a `ptr: *mut opendal_bytes`.

Because `bs: opendal_bytes` doesn't "moved" the struct from C. Instead, it makes a copy.

Users should set `bs.data = NULL` and `bs.len = 0` etc. outside.

Reply:
Updated.